File: src/viewportMeta.ts
/**
 * viewportMeta.ts — viewport meta tag detector. T76.
 *
 * Three findings:
 *
 *   - viewport.missing                strict
 *     No `<meta name="viewport" ...>` at all. The page renders at
 *     the desktop default of 980px and is then scaled by the
 *     mobile browser, producing the "tiny everything" experience
 *     that has plagued legacy sites since 2008. WCAG 1.4.10
 *     (Reflow, AA) requires content to be presented without
 *     two-dimensional scrolling — a missing viewport tag breaks
 *     this for almost every screen narrower than 980px.
 *
 *   - viewport.no-device-width        strict
 *     Tag exists but content lacks `width=device-width`. Same
 *     symptom as missing tag for most pages — the browser uses
 *     the default 980px viewport width.
 *
 *   - viewport.zoom-disabled          strict
 *     Content includes `user-scalable=no`, `user-scalable=0`, or
 *     `maximum-scale` ≤ 1. Users with low vision cannot pinch-zoom.
 *     WCAG 1.4.4 (Resize text, AA) — explicitly forbidden by the
 *     spec since 2018. iOS Safari started ignoring it for the
 *     same reason; Android Chrome still honors it. Even if iOS
 *     bypasses it, the AUTHOR INTENT is hostile and the page
 *     should be fixed.
 *
 * Mirror: crates/crawler-detectors/src/viewport_meta.rs.
 */
import type { Page } from 'playwright';

export interface ViewportMetaFinding {
  severity: 'strict' | 'warn';
  kind: string;
  detail: string;
  evidence: Record<string, unknown>;
}

export interface ViewportMetaSnapshot {
  pageUrl: string;
  /** True iff the page has at least one <meta name="viewport"> in head. */
  present: boolean;
  /** Raw `content` attribute of the first viewport meta, '' if absent. */
  content: string;
}

export async function captureViewportMetaSnapshot(
  page: Page,
): Promise<ViewportMetaSnapshot> {
  const pageUrl = page.url();
  const evalFn = `(() => {
    const m = document.querySelector('head meta[name="viewport"]');
    if (!m) return { present: false, content: '' };
    const c = m.getAttribute('content') || '';
    return { present: true, content: c };
  })()`;
  const result = (await page.evaluate(evalFn)) as {
    present: boolean;
    content: string;
  };
  return { pageUrl, present: result.present, content: result.content };
}

/**
 * Parse a viewport `content` string into a key→value map.
 *
 * `width=device-width, initial-scale=1, user-scalable=no` →
 *   { width: 'device-width', 'initial-scale': '1', 'user-scalable': 'no' }
 *
 * Defensive: tolerates extra whitespace, trailing commas, missing
 * value (key with no `=`), and case differences in keys (per spec
 * keys are case-insensitive). Values are trimmed but case-preserved
 * since some are numeric and case doesn't matter.
 */
export function parseViewportContent(content: string): Record<string, string> {
  const out: Record<string, string> = {};
  for (const segment of content.split(/[,;]/)) {
    const trimmed = segment.trim();
    if (!trimmed) continue;
    const eq = trimmed.indexOf('=');
    if (eq < 0) {
      out[trimmed.toLowerCase()] = '';
      continue;
    }
    const k = trimmed.slice(0, eq).trim().toLowerCase();
    const v = trimmed.slice(eq + 1).trim();
    if (k) out[k] = v;
  }
  return out;
}

export function detectViewportMetaIssues(
  snap: ViewportMetaSnapshot,
): ViewportMetaFinding[] {
  const out: ViewportMetaFinding[] = [];

  if (!snap.present) {
    out.push({
      severity: 'strict',
      kind: 'viewport.missing',
      detail: `No <meta name="viewport"> in the page <head>. Mobile browsers will render at the legacy 980px default viewport then scale down — text is unreadable, tap targets are unreachable, and WCAG 1.4.10 (Reflow, AA) fails. Add <meta name="viewport" content="width=device-width, initial-scale=1">.`,
      evidence: { present: false },
    });
    return out;
  }

  const parts = parseViewportContent(snap.content);

  if ((parts['width'] || '').toLowerCase() !== 'device-width') {
    out.push({
      severity: 'strict',
      kind: 'viewport.no-device-width',
      detail: `<meta name="viewport"> is present but content lacks 'width=device-width'. Mobile browsers fall back to the legacy 980px default. Got: '${snap.content}'. Fix: 'width=device-width, initial-scale=1'.`,
      evidence: { content: snap.content, parsed: parts },
    });
  }

  // Zoom-disabling forms.
  const userScalable = (parts['user-scalable'] || '').toLowerCase();
  const maxScaleStr = parts['maximum-scale'] || '';
  const maxScale = parseFloat(maxScaleStr);
  const zoomDisabled =
    userScalable === 'no' ||
    userScalable === '0' ||
    (Number.isFinite(maxScale) && maxScale <= 1);

  if (zoomDisabled) {
    out.push({
      severity: 'strict',
      kind: 'viewport.zoom-disabled',
      detail: `<meta name="viewport"> disables user pinch-zoom (user-scalable='${userScalable}', maximum-scale='${maxScaleStr}'). WCAG 1.4.4 (Resize text, AA): users must be able to scale text up to 200%. Hostile to low-vision users. Remove user-scalable=no and any maximum-scale ≤ 1.`,
      evidence: { content: snap.content, userScalable, maximumScale: maxScaleStr },
    });
  }

  return out;
}

export async function checkViewportMeta(
  page: Page,
): Promise<{ snapshot: ViewportMetaSnapshot; findings: ViewportMetaFinding[] }> {
  const snapshot = await captureViewportMetaSnapshot(page);
  const findings = detectViewportMetaIssues(snapshot);
  return { snapshot, findings };
}

File: crates/crawler-detectors/src/viewport_meta.rs
//! `viewport_meta` — viewport meta tag detector.
//!
//! Three findings — all strict (each is a real, blocking
//! mobile/a11y defect):
//!
//!   * `viewport.missing`           no `<meta name="viewport">`
//!   * `viewport.no-device-width`   tag present but no `width=device-width`
//!   * `viewport.zoom-disabled`     `user-scalable=no|0` or
//!                                  `maximum-scale ≤ 1`
//!
//! Mirrors `src/viewportMeta.ts` — JS string + finding kinds +
//! severities byte-equivalent.
//!
//! AVP-2 INVARIANTS
//! ----------------
//! * `unsafe_code = "deny"`.
//! * Pure detector function; no I/O.

use serde::{Deserialize, Serialize};

/// Page-side eval. Returns the first viewport meta tag's content
/// attribute (or `present=false` if no such tag).
pub const VIEWPORT_META_JS: &str = r##"(() => {
    const m = document.querySelector('head meta[name="viewport"]');
    if (!m) return { present: false, content: '' };
    const c = m.getAttribute('content') || '';
    return { present: true, content: c };
})()"##;

/// Snapshot.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct ViewportMetaSnapshot {
    /// Page URL at capture time.
    pub page_url: String,
    /// True iff at least one `<meta name="viewport">` exists in head.
    pub present: bool,
    /// Raw `content` attribute of the first viewport meta tag, empty
    /// if `present == false`.
    pub content: String,
}

/// Parse a viewport `content` string into a key→value map.
///
/// `width=device-width, initial-scale=1, user-scalable=no` →
///   `{"width": "device-width", "initial-scale": "1", "user-scalable": "no"}`
///
/// Tolerant of whitespace, trailing separators, and flag-style
/// tokens (a key with no `=` maps to empty-string value). Keys are
/// lowercased per spec; values are case-preserved (most are numeric).
#[must_use]
pub fn parse_viewport_content(content: &str) -> std::collections::BTreeMap<String, String> {
    let mut out = std::collections::BTreeMap::new();
    for segment in content.split([',', ';']) {
        let trimmed = segment.trim();
        if trimmed.is_empty() {
            continue;
        }
        if let Some(eq) = trimmed.find('=') {
            let k = trimmed[..eq].trim().to_lowercase();
            let v = trimmed[eq + 1..].trim().to_owned();
            if !k.is_empty() {
                out.insert(k, v);
            }
        } else {
            out.insert(trimmed.to_lowercase(), String::new());
        }
    }
    out
}

/// Apply detection rules. Pure function. Mirror of
/// `detectViewportMetaIssues` in `src/viewportMeta.ts`.
#[must_use]
pub fn detect_viewport_meta_issues(snap: &ViewportMetaSnapshot) -> Vec<crate::AxisFinding> {
    let mut out = Vec::<crate::AxisFinding>::new();

    if !snap.present {
        out.push(crate::AxisFinding {
            severity: crate::AxisSeverity::Strict,
            kind: "viewport.missing".to_owned(),
            detail: "No <meta name=\"viewport\"> in the page <head>. Mobile browsers will render at the legacy 980px default viewport then scale down — text is unreadable, tap targets are unreachable, and WCAG 1.4.10 (Reflow, AA) fails. Add <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">.".to_owned(),
        });
        return out;
    }

    let parts = parse_viewport_content(&snap.content);

    let width_ok = parts
        .get("width")
        .map(|v| v.eq_ignore_ascii_case("device-width"))
        .unwrap_or(false);
    if !width_ok {
        out.push(crate::AxisFinding {
            severity: crate::AxisSeverity::Strict,
            kind: "viewport.no-device-width".to_owned(),
            detail: format!(
                "<meta name=\"viewport\"> is present but content lacks 'width=device-width'. Mobile browsers fall back to the legacy 980px default. Got: '{}'. Fix: 'width=device-width, initial-scale=1'.",
                snap.content
            ),
        });
    }

    let user_scalable = parts.get("user-scalable").map(|s| s.to_lowercase()).unwrap_or_default();
    let max_scale_str = parts.get("maximum-scale").cloned().unwrap_or_default();
    let max_scale: Option<f64> = max_scale_str.parse().ok();
    let zoom_disabled = user_scalable == "no"
        || user_scalable == "0"
        || max_scale.map(|v| v <= 1.0).unwrap_or(false);

    if zoom_disabled {
        out.push(crate::AxisFinding {
            severity: crate::AxisSeverity::Strict,
            kind: "viewport.zoom-disabled".to_owned(),
            detail: format!(
                "<meta name=\"viewport\"> disables user pinch-zoom (user-scalable='{user_scalable}', maximum-scale='{max_scale_str}'). WCAG 1.4.4 (Resize text, AA): users must be able to scale text up to 200%. Hostile to low-vision users. Remove user-scalable=no and any maximum-scale ≤ 1."
            ),
        });
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn snap_with(content: &str) -> ViewportMetaSnapshot {
        ViewportMetaSnapshot {
            page_url: "http://t/".to_owned(),
            present: true,
            content: content.to_owned(),
        }
    }

    fn snap_missing() -> ViewportMetaSnapshot {
        ViewportMetaSnapshot {
            page_url: "http://t/".to_owned(),
            present: false,
            content: String::new(),
        }
    }

    #[test]
    fn js_brackets_balanced() {
        assert_eq!(VIEWPORT_META_JS.matches('(').count(), VIEWPORT_META_JS.matches(')').count());
        assert_eq!(VIEWPORT_META_JS.matches('{').count(), VIEWPORT_META_JS.matches('}').count());
    }

    #[test]
    fn clean_no_findings() {
        let f = detect_viewport_meta_issues(&snap_with("width=device-width, initial-scale=1"));
        assert!(f.is_empty(), "{:?}", f);
    }

    #[test]
    fn missing_tag_fires_strict() {
        let f = detect_viewport_meta_issues(&snap_missing());
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "viewport.missing");
        assert_eq!(f[0].severity, crate::AxisSeverity::Strict);
    }

    #[test]
    fn no_device_width_strict() {
        let f = detect_viewport_meta_issues(&snap_with("initial-scale=1"));
        assert!(f.iter().any(|x| x.kind == "viewport.no-device-width"));
    }

    #[test]
    fn user_scalable_no_strict() {
        let f = detect_viewport_meta_issues(&snap_with(
            "width=device-width, user-scalable=no",
        ));
        assert!(f.iter().any(|x| x.kind == "viewport.zoom-disabled"));
    }

    #[test]
    fn user_scalable_zero_strict() {
        let f = detect_viewport_meta_issues(&snap_with(
            "width=device-width, user-scalable=0",
        ));
        assert!(f.iter().any(|x| x.kind == "viewport.zoom-disabled"));
    }

    #[test]
    fn max_scale_one_strict() {
        let f = detect_viewport_meta_issues(&snap_with(
            "width=device-width, maximum-scale=1",
        ));
        assert!(f.iter().any(|x| x.kind == "viewport.zoom-disabled"));
    }

    #[test]
    fn max_scale_below_one_strict() {
        let f = detect_viewport_meta_issues(&snap_with(
            "width=device-width, maximum-scale=0.9",
        ));
        assert!(f.iter().any(|x| x.kind == "viewport.zoom-disabled"));
    }

    #[test]
    fn max_scale_two_passes() {
        let f = detect_viewport_meta_issues(&snap_with(
            "width=device-width, maximum-scale=2",
        ));
        assert!(!f.iter().any(|x| x.kind == "viewport.zoom-disabled"));
    }

    #[test]
    fn whitespace_tolerated() {
        let f = detect_viewport_meta_issues(&snap_with(
            "  width = device-width ,  initial-scale = 1 , ",
        ));
        assert!(f.is_empty(), "{:?}", f);
    }

    #[test]
    fn keys_case_insensitive() {
        let f = detect_viewport_meta_issues(&snap_with("WIDTH=device-width"));
        assert!(f.is_empty(), "{:?}", f);
    }

    #[test]
    fn parser_flag_token_no_eq() {
        let p = parse_viewport_content("foo, width=device-width");
        assert_eq!(p.get("foo").map(String::as_str), Some(""));
        assert_eq!(p.get("width").map(String::as_str), Some("device-width"));
    }

    #[test]
    fn combined_breakage_two_findings() {
        let f = detect_viewport_meta_issues(&snap_with(
            "initial-scale=1, user-scalable=no",
        ));
        assert!(f.iter().any(|x| x.kind == "viewport.no-device-width"));
        assert!(f.iter().any(|x| x.kind == "viewport.zoom-disabled"));
    }
}
